src/tox/session/env_select.py

```diff
@@ -389,6 +391,16 @@ def _mark_provision(self, on: bool, provision_tox_env: str) -> None:  # noqa: FB
         self._provision = on, provision_tox_env
 
 
+def _find_env_match(value: str, state_conf: set) -> bool:
+    return any(value in conf.split("-") for conf in state_conf) or any(
+        value == "-".join(combo) for combo in permutations(state_conf, 2)
+    )
+
+
+def _is_valid_exception(env: str) -> bool:
+    return env.startswith("py") or env in (".pkg",)
```

Comment:
Can you ensure that 3.9/3.11 is also handled as an exception, we can use a Regen for that 👍

Reply:
Mind clarifying? What use-case is currently not working?

Reply:
`tox r -e 3.11` and `tox r -e 3.11-magic` should still work 🤔

Reply:
Something like this: https://regex101.com/r/u2Kwuy/1 ? @gaborbernat

Reply:
https://www.debuggex.com/r/8ZDSfICtm-vkvTjs was my approach

Reply:
Seems more complex and don't cover the "-magic" case?

Is my approach missing a case?

Reply:
Well the -magic doesn't matter. If any of the factors match this regex then we should accept the whole env name.

Reply:
The idea here is to still allow dynamic environments for envs having my regex, plus any other factors (and their combinations) the user defined.

Reply:
Ok, but none of our regexs would match `test-cov` or `cov-test`, right? 🤔 

It's almost 4AM where I'm at. I think I need to look at this with fresh eyes.

However, I think generally explicit > implicit. So maybe this is just introducing complexity that is unnecessary. 🤷‍♂️

Reply:
Can you show an example tox.ini that creates a environment with three or more factors _without_ having `env_list` contain any factors.

e.g.:

```ini
[tox]
env_list=3.9
commands =
  cov: python -c 'print("with cov")'
 !cov: python -c 'print("without cov")'
```

Example above will create a maximum of two possible factors, right?

`3.9` and `3.9-cov`

How can that be extended to three or more factors?